### PR TITLE
feat: allow sending headers upstream returned by OPA server

### LINKED ISSUE #9704: feat:  opa plugins, add the ability to set headers
### Description

opa plugin can set headers if the policy response is false - https://apisix.apache.org/docs/apisix/plugins/opa/#using-custom-response 
I would like to be able to set headers for a request if the policy response is true. 
Use cases - to send to the backend the status of the policy, the policy name, fields from the JWT token.
The changes are not so big, so I can prepare a MR

### PR DESCRIPTION
**Pull request by Revolyssup**
### Description

Fixes #9704 

Documentation to be updated
### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by monkeyDluffy6017** [comment]
Please make the ci pass

**Comment by Revolyssup** [comment]
> Please make the ci pass and no test cases?

Added test case @monkeyDluffy6017

**Review by lingsamuel — APPROVED** [review]

**Comment by Revolyssup** [comment]
@monkeyDluffy6017 PTAL

File: apisix/plugins/opa.lua [hunk 1]
```diff
@@ -125,6 +127,11 @@ function _M.access(conf, ctx)
         end
 
         return status_code, reason
+    else if conf.send_header_upstream and result.headers then
+           for key,value in pairs(result.headers) do
+                core.request.set_header(ctx,key,value)
+           end
+        end
```
Comment (monkeyDluffy6017) [inline apisix/plugins/opa.lua:134]: Is it necessary to proxy all of opa's request headers?

Comment (Revolyssup) [reply apisix/plugins/opa.lua:134]: What else do you propose? How to select which ones to proxy?

**Comment by Sn0rt** [comment]
@chusov pls help us take a look

File: apisix/plugins/opa.lua (hunk 1 above)
Comment (Sn0rt) [reply apisix/plugins/opa.lua:134]: what HTTP header should pass upstream? instead of forwarding all HTTP header

**Comment by chusov** [comment]
Maybe add a parameter with a list of headers? 
`send_header_upstream_list`

**Comment by Revolyssup** [comment]
> Maybe add a parameter with a list of headers? `send_header_upstream_list`

@monkeyDluffy6017 What do you think about this?

**Comment by monkeyDluffy6017** [comment]
@chusov i think the headers that need to be passed to upstream from OPA server should be fixed, what headers do you exactly need?

**Comment by chusov** [comment]
@monkeyDluffy6017 
at the moment it is enough X-Username, in the future there were plans to add more depending on the service

**Comment by monkeyDluffy6017** [comment]
> > Maybe add a parameter with a list of headers? `send_header_upstream_list`
> 
> @monkeyDluffy6017 What do you think about this?

LGTM

**Review by Sn0rt — APPROVED** [review]

**Comment by Revolyssup** [comment]
> > > Maybe add a parameter with a list of headers? `send_header_upstream_list`
> > 
> > 
> > @monkeyDluffy6017 What do you think about this?
> 
> LGTM

Okay I'm making the change. @Sn0rt This one change remaining as suggested above by user

File: apisix/plugins/opa.lua [hunk 2]
```diff
@@ -20,6 +20,7 @@ local http   = require("resty.http")
 local helper = require("apisix.plugins.opa.helper")
 local type   = type
 
+local pairs = pairs
```
Comment (monkeyDluffy6017) [inline apisix/plugins/opa.lua:23]: Why not put this under this line `local type = type`

Comment (monkeyDluffy6017) [reply apisix/plugins/opa.lua:22]: And a blank is needed before the line 23

File: apisix/plugins/opa.lua [hunk 3]
```diff
@@ -37,6 +37,14 @@ local schema = {
...
         },
         keepalive = {type = "boolean", default = true},
+        send_headers_upstream = {
+            type = "array",
+            minItems = 0,
+            items = {
+                type = "string"
+            },
+            description = "list of headers to pass to upstream in request"
+        },
```
Comment (monkeyDluffy6017) [inline apisix/plugins/opa.lua:47]:
```suggestion
        send_headers_upstream = {
            type = "array",
            minItems = 1,
            items = {
                type = "string"
            },
            description = "list of headers to pass to upstream in request"
        },
```

File: apisix/plugins/opa.lua [hunk 4]
```diff
@@ -125,6 +133,18 @@ function _M.access(conf, ctx)
...
+            for key in pairs(conf.send_headers_upstream) do
+                headersToSend[key] = true
+            end
+           for key,value in pairs(result.headers) do
+                if headersToSend[key] then
+                    core.request.set_header(ctx,key,value)
+                end
+
+           end
+        end
```
Comment (monkeyDluffy6017) [inline apisix/plugins/opa.lua:147]:
```suggestion
    else if result.headers and conf.send_headers_upstream then
            local headersToSend = {}
            for key in pairs(conf.send_headers_upstream) do
                headersToSend[key] = true
            end
           for key,value in pairs(result.headers) do
                if headersToSend[key] then
                    core.request.set_header(ctx,key,value)
                end

           end
        end
```

Comment (monkeyDluffy6017) [reply apisix/plugins/opa.lua:147]: The indent is not right, only four spaces are needed

Comment (Revolyssup) [reply apisix/plugins/opa.lua:147]: done

File: apisix/plugins/opa.lua (hunk 3 above)
Comment (Revolyssup) [reply apisix/plugins/opa.lua:47]: done

File: apisix/plugins/opa.lua (hunk 2 above)
Comment (Revolyssup) [reply apisix/plugins/opa.lua:22]: done

File: apisix/plugins/opa.lua [hunk 5]
```diff
@@ -125,6 +134,18 @@ function _M.access(conf, ctx)
         end
 
         return status_code, reason
+    else if result.headers and conf.send_headers_upstream then
+        local headersToSend = {}
+        for key in pairs(conf.send_headers_upstream) do
+            headersToSend[key] = true
+        end
+        for key,value in pairs(result.headers) do
```
Comment (monkeyDluffy6017) [inline apisix/plugins/opa.lua:142]:
```suggestion
        for key, value in pairs(result.headers) do
```

File: apisix/plugins/opa.lua [hunk 6]
```diff
@@ -125,6 +134,18 @@ function _M.access(conf, ctx)
...
+            if headersToSend[key] then
+                core.request.set_header(ctx,key,value)
+            end
+
```
Comment (monkeyDluffy6017) [inline apisix/plugins/opa.lua:146]: No blank is needed

File: apisix/plugins/opa.lua [hunk 7]
```diff
@@ -125,6 +134,18 @@ function _M.access(conf, ctx)
...
 
         return status_code, reason
+    else if result.headers and conf.send_headers_upstream then
+        local headersToSend = {}
+        for key in pairs(conf.send_headers_upstream) do
+            headersToSend[key] = true
+        end
+        for key,value in pairs(result.headers) do
+            if headersToSend[key] then
+                core.request.set_header(ctx,key,value)
```
[the remark this answers is not in the document]
Comment (monkeyDluffy6017) [reply apisix/plugins/opa.lua:144]: I think you could refactor this block like this:
```
        for key in pairs(conf.send_headers_upstream) do
            if result.headers[key] then
                core.request.set_header(ctx, key, value)
            end
        end
```

[the remark this answers is not in the document]
Comment (Revolyssup) [reply apisix/plugins/opa.lua:144]: Makes much sense. Done

**Review by leslie-tsang — APPROVED** [review]

**Pull request merged**